src/databricks/labs/lsql/backends.py

```diff
@@ -135,20 +135,29 @@ def execute(self, sql: str, *, catalog: str | None = None, schema: str | None =
     def fetch(self, sql: str, *, catalog: str | None = None, schema: str | None = None) -> Iterator[Any]:
         raise NotImplementedError
 
-    def save_table(self, full_name: str, rows: Sequence[DataclassInstance], klass: Dataclass, mode="append"):
+    def save_table(
+        self,
+        full_name: str,
+        rows: Sequence[DataclassInstance],
+        klass: Dataclass,
+        mode: Literal["append", "overwrite"] = "append",
+    ):
         rows = self._filter_none_rows(rows, klass)
         self.create_table(full_name, klass)
-        if len(rows) == 0:
+        if not rows:
+            if mode == "overwrite":
+                self.execute(f"TRUNCATE TABLE {full_name}")
```

Comment:
```suggestion
                self.execute(f"DELETE FROM {full_name}")
```
otherwise we remove delta log and can't recover data. isn't `INSERT OVERWRITE ... VALUES ()` doing the same?

Reply:
Using `TRUNCATE` does not remove the Delta history:
```sql
create table test_truncate as select * from values 1, 2, 3;
insert into test_truncate values 4, 5, 6;

truncate table test_truncate;

select version, timestamp, operation, operationMetrics, engineInfo
from (describe history test_truncate);
```
| `version` | `timestamp` | `operation` | `operationMetrics` | `engineInfo` |
| --------- | ----------- | ----------- | ------------------ | ------------ |
| 2 | 2024-09-30 09:20:35.000000 | TRUNCATE | `{"numRemovedFiles":"2","numDeletedRows":"6","executionTimeMs":"15","numRemovedBytes":"906"}` | Databricks-Runtime/15.3.x-photon-scala2.12 |
| 1 | 2024-09-30 09:20:34.000000 | WRITE | `{"numFiles":"1","numOutputRows":"3","numOutputBytes":"453"}` | Databricks-Runtime/15.3.x-photon-scala2.12 |
| 0 | 2024-09-30 09:20:33.000000 | CREATE TABLE AS SELECT | `{"numFiles":"1","numOutputRows":"3","numOutputBytes":"453"}` | Databricks-Runtime/15.3.x-photon-scala2.12 |

The `INSERT OVERWRITE` operation would be semantically equivalent, but this is part of the block that handles `.save_table()` when no rows are provided: there was an (existing) optimisation in place to turn this into a no-op by returning early. This was correct for append-mode, but not really what would be expected in overwrite mode. It still returns early.